include/rive/renderer.hpp

```diff
@@ -16,27 +17,66 @@ namespace rive
 {
 	class Vec2D;
 
+    using ColorInt = uint32_t;
+
 	enum class RenderPaintStyle
 	{
 		stroke,
 		fill
 	};
 
+    enum class RenderTileMode {
+        clamp,
+        repeat,
+        mirror,
+        decal,  // fill outside the domain with transparent
+    };
+
+    /*
+     *  Base class for Render objects that specify the src colors.
+     *
+     *  Shaders are immutable, and sharable between multiple paints, etc.
+     *
+     *  It is common that a shader may be created with a 'localMatrix'. If this is
+     *  not null, then it is applied to the shader's domain before the Renderer's CTM.
+     */
+    class RenderShader : public RefCnt {};
+
+    extern rcp<RenderShader> makeLinearGradient(float sx, float sy,
+                                                float ex, float ey,
+                                                const ColorInt colors[],    // [count]
+                                                const float stops[],        // [count]
+                                                int count,
+                                                RenderTileMode,
+                                                const Mat2D* localMatrix = nullptr);
+
+    extern rcp<RenderShader> makeRadialGradient(float cx, float cy, float radius,
+                                                const ColorInt colors[],    // [count]
+                                                const float stops[],        // [count]
+                                                int count,
+                                                RenderTileMode,
+                                                const Mat2D* localMatrix = nullptr);
+
+    extern rcp<RenderShader> makeSweepGradient(float cx, float cy,
+                                               const ColorInt colors[],    // [count]
+                                               const float stops[],        // [count]
+                                               int count,
+                                               const Mat2D* localMatrix = nullptr);
+
 	class RenderPaint
 	{
 	public:
 		virtual void style(RenderPaintStyle style) = 0;
-		virtual void color(unsigned int value) = 0;
+        // This is used only if the 'shader' is null.
+		virtual void color(ColorInt value) = 0;
 		virtual void thickness(float value) = 0;
 		virtual void join(StrokeJoin value) = 0;
 		virtual void cap(StrokeCap value) = 0;
 		virtual void blendMode(BlendMode value) = 0;
+        // If shader is not null, then ignore the 'color'
+        virtual void shader(rcp<RenderShader>) = 0;
```

Comment:
This seems sound!

It makes me wonder if there's a world where all of these "complications" of the pixel fill (not fill/stroke) model can be generalized to shader specializations. You want a Stroke, set the Shader to a StrokingShader that has thickness, join, cap properties. It gets messy with interdependencies, see a bit of how we solved this via composition in Rive's Core guts:

https://www.notion.so/Shape-Painting-90808853fe1f48339e3b12dbf01db19e

Reply:
Not that link solves any of this, but it separates Fill/Stroke such that Color/Gradient can be added to each without caring if it's a fill or a stroke. This is way more complex here, and we did originally model this rendering abstraction to work nicely with Skia and Context2D, but if we weren't, would we do things differently?